Build Goose in a Docker Container

ZhenLian opened the pull request:
## Description
This update partially addresses [issue #1072](https://github.com/block/goose/issues/1072), where users requested the ability to build and run Goose in a Docker container. Since users may have different use cases for using Docker, I've added a [discussion thread](https://github.com/block/goose/discussions/1496) to explore additional use cases and allow users to contribute their own. This update also includes a brief guide on how to build Goose within a Docker container.

## Testing
I tested this on my local laptop. 

1. When running the Goose CLI in the docker container, you can observe the following:
![goose1](https://github.com/user-attachments/assets/12c4b1fe-922f-490a-8ebe-e057378f4b9b)

2. On the first run of goose config, you select the API key settings:
![goose2](https://github.com/user-attachments/assets/032f0f48-6c9b-47c7-b470-c9d728a34a26)

3. On the second run of goose config, you choose the extension settings (I selected the developer tool).
![goose3](https://github.com/user-attachments/assets/0c618950-2acf-4088-b909-141880ec32ff)

4. Running goose session will then initiate a Goose session with developer tool capabilities.
![goose4](https://github.com/user-attachments/assets/9951bfe3-7f1f-422b-b912-bc246e4d9767)

lily-de reviewed (approved):
thanks for the documentation! 🙏

yingjiehe-xyz commented:
Thanks for the contribution! Have you tried run goose with custom extensions?

yingjiehe-xyz commented on Dockerfile:
```diff
@@ -0,0 +1,123 @@
+    rm -rf /var/lib/apt/lists/*
+
+# Install common development tools
+RUN apt-get update && apt-get install -y \
```
do we want to install uv as well?

ZhenLian replied:
Certainly! I included a line to install uv and verified within the Docker container that it is properly installed.

yingjiehe-xyz commented on docker-compose.yml:
```diff
@@ -0,0 +1,31 @@
+      # Preserve git author info
+      - GIT_AUTHOR_NAME=${GIT_AUTHOR_NAME:-Goose User}
+      - GIT_AUTHOR_EMAIL=${GIT_AUTHOR_EMAIL:-goose@example.com}
+      # Set GOOGLE_API_KEY to your own API key
```
should we ignore the env variable here? I think it would be better to set it via goose configure

ZhenLian replied:
Yeah, there are some nuances here. Docker operates in a headless environment, and it lacks certain built-in packages for user display. When we attempt to set and save the key into the keyring using `goose configure`, we will encounter this error:
```
Error Failed to access secure storage: Platform secure storage failure: DBus error: Unable to autolaunch a dbus-daemon without a $DISPLAY for X11 
  Please check your system's secure storage and run 'goose configure' again. 
  If your system is unable to use secure storage, please try setting secret key(s) via environment variables.
```
In the Dockerfile, I've included a script to try installing DBus, but it hasn't worked yet. I'll continue monitoring this issue, and if I find a solution for storing this in the keyring, I'll update the code and instructions accordingly!

kalvinnchau commented on docker-compose.yml:
```diff
@@ -0,0 +1,31 @@
+    volumes:
+      # Mount user's directory with read-write access
+      - .:/home/goose/workspace
+      # Mount git config (optional)
+      - ~/.gitconfig:/home/goose/.gitconfig:ro
+      # Mount SSH keys (optional)
+      - ~/.ssh:/home/goose/.ssh:ro
```
how are these defined as optional? wouldn't this always mount the directories via docker compose?

ZhenLian replied:
Yes, my goal was to inform users that the volume mounting configurations are adjustable - they can decide whether to comment out those lines. That seems to be confusing though. I removed the optional wording, as users can simply read the code and remove those lines if necessary.

michaelneale reviewed (changes requested):
the dockerfile and compose should be in docs dir, not root, as they are for illustration purposes is that right?

ZhenLian commented on Dockerfile:
```diff
@@ -0,0 +1,123 @@
+# Build stage
+FROM rust:1.75-bullseye AS builder
```
That's a good catch! Version 1.75 was initially selected when I asked Cloude to create a template, but I didn't change it afterward :) I've now updated the line to `FROM rust:bullseye AS builder`

ZhenLian commented on docker-compose.yml:
```diff
@@ -0,0 +1,31 @@
+      - SSH_AUTH_SOCK
+    stdin_open: true
+    tty: true
+    entrypoint: ["/bin/bash"]
```
Good catch! The `goose-wrapper` entrypoint isn't being utilized. I've removed it and revised some of the wording.

ZhenLian commented:
> the dockerfile and compose should be in docs dir, not root, as they are for illustration purposes is that right?

Sure, I moved them to the docs folder and updated the instructions for how to compile them.

ZhenLian commented:
> Have you tried run goose with custom extensions?

Not yet. My goal was to create a basic Docker template file to help users understand how to build Docker images from it. Once they're connected to the container, they have the freedom to add additional dependencies or custom extensions as they wish. 
That being said, if there are any particularly useful extensions, I'd be glad to test them and add them to the documentation (so far, I've only worked with the built-in extensions :)). Given that Docker runs in a headless environment, some extensions might not be as advantageous as others. Maybe we could wait until a user specifically requests an extension?

ZhenLian commented:
@kalvinnchau @michaelneale @yingjiehe-xyz 
Cool, I think I've addressed all your comments above. Could I ask for another review of this PR? Thank you very much!

michaelneale reviewed (approved):
thanks!

ZhenLian commented:
@yingjiehe-xyz @michaelneale Awesome, thanks for the review!
@kalvinnchau I am merging this, but if you happen to spot anything in the future, feel free to let me know - happy to submit a fix in a supplemental PR after this!

evangineer commented:
I'm working on integrating Goose with a multi-Docker Compose setup, and have been puzzled by the line below. 

`- OOSE_HOME=/root/.goose`

Is it a typo and why can't I find any info at all for either `OOSE_HOME` or `GOOSE_HOME` in the docs, the repo and the discord?
https://github.com/block/goose/blob/b1bf9d7173aea9c41cc59ac9114801e709719792/documentation/docs/docker/docker-compose.yml#L18

blevz commented:
Yup, this satisfies the thrust of my request.  Tysm
